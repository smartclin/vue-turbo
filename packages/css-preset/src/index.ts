import type { Preset, UserConfig } from '@unocss/core'
import type { UnocssNuxtOptions } from '@unocss/nuxt'
import type { PresetWindOptions, Theme } from '@unocss/preset-wind'
import { presetIcons } from '@unocss/preset-icons'
import { presetWind, transformerDirectives, transformerVariantGroup } from 'unocss'
import { rules } from './rules'
import { theme } from './theme'

// @unocss-include

/**
 * Preset Vue Turbo for UnoCSS
 */
export function presetVueTurbo(): Preset<Theme> {
  return {
    name: 'unocss-preset-vue-turbo',
    rules,
    safelist: [
      'hidden',
    ],
    preflights: [
      {
        layer: 'basestyles',
        getCSS: () => `
          html,
          body,
          #app {
            height: 100%;
            margin: 0;
            padding: 0;
          }

          html.dark {
            background: #121212;
            color-scheme: dark;
          }

          :root {
            color-scheme: light dark;
            color: rgba(255, 255, 255, 0.87);
            background-color: #242424;
          }

          @media (prefers-color-scheme: light) {
            :root {
              color: #213547;
              background-color: #ffffff;
            }
            a:hover {
              color: #747bff;
            }
          }
        `,
      },
    ],
    shortcuts: [],
    theme,
  }
}

interface CustomOptions {
  /**
   * Custom options for PresetWind
   */
  customPresetWindOptions?: PresetWindOptions
}
interface CustomNuxtConfig extends UnocssNuxtOptions, CustomOptions {}

/**
 * Extends unocss/vite Plugin Options Config
 */
export function extendUnocssOptions({ customPresetWindOptions, ...options }: CustomNuxtConfig = {}): UserConfig {
  return {
    ...options,
    presets: [
      presetWind({
        ...(customPresetWindOptions || {}),
      }),
      presetVueTurbo(),
      ...(options.presets || []),
      presetIcons({
        extraProperties: {
          display: 'inline-block',
        },
      }),
    ],
    transformers: [
      transformerDirectives(),
      transformerVariantGroup(),
      ...(options.transformers || []),
    ],
    content: {
      pipeline: {
        include: [
          '**.ts',
          '**.vue',
          '**.mdx',
        ],
      },
    },
  }
}
