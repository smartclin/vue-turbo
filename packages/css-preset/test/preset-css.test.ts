import { createGenerator, escapeSelector } from '@unocss/core'
import { presetIcons } from '@unocss/preset-icons'
import { presetWind } from '@unocss/preset-wind'
import { describe, expect, it } from 'vitest'

import { presetVueTurbo } from '../src'

import { presetCssNonTargets, presetCssTargets } from './assets/preset-css-preset-targets'

const uno = createGenerator({
  ...presetVueTurbo(),
  presets: [
    presetWind(),
    presetIcons({
      extraProperties: {
        display: 'inline-block',
      },
    }),
  ],
  safelist: [],
})

describe('test css-preset', () => {
  it('targets', async () => {
    const targets = presetCssTargets
    const code = targets.join(' ')
    const { css } = await uno.generate(code)
    const { css: css2 } = await uno.generate(code)

    const unmatched = []
    for (const i of targets) {
      if (!css.includes(escapeSelector(i)))
        unmatched.push(i)
    }

    expect(unmatched).toEqual([])
    expect(css).toMatchSnapshot()
    expect(css).toEqual(css2)
  })

  it('non-targets', async () => {
    const { matched } = await uno.generate(new Set(presetCssNonTargets), { preflights: false })

    expect([...matched]).toEqual([])
  })

  it('containers', async () => {
    const targets = [
      'container',
      'md:container',
      'lg:container',
    ]
    const nonTargets = [
      '__container',
    ]
    const { css, matched } = await uno.generate(new Set([...targets, ...nonTargets]), { preflights: false })

    expect(matched).toEqual(new Set(targets))
    expect(css).toMatchSnapshot()
  })
})
