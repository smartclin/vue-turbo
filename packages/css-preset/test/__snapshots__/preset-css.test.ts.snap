// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test css-preset > containers 1`] = `
"/* layer: shortcuts */
.container{width:100%;}
@media (min-width: 640px){
.container{max-width:640px;}
}
@media (min-width: 768px){
.container,
.md\\:container{max-width:768px;}
}
@media (min-width: 1024px){
.container,
.lg\\:container,
.md\\:container{max-width:1024px;}
}
@media (min-width: 1280px){
.container,
.lg\\:container,
.md\\:container{max-width:1280px;}
}
@media (min-width: 1536px){
.container,
.lg\\:container,
.md\\:container{max-width:1536px;}
}"
`;

exports[`test css-preset > targets 1`] = `
"/* layer: preflights */
*,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgb(0 0 0 / 0);--un-ring-shadow:0 0 rgb(0 0 0 / 0);--un-shadow-inset: ;--un-shadow:0 0 rgb(0 0 0 / 0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgb(147 197 253 / 0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgb(0 0 0 / 0);--un-ring-shadow:0 0 rgb(0 0 0 / 0);--un-shadow-inset: ;--un-shadow:0 0 rgb(0 0 0 / 0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgb(147 197 253 / 0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}
/* layer: icons */
.i-system-uicons-contract{--un-icon:url("data:image/svg+xml;utf8,%3Csvg viewBox='0 0 21 21' display='inline-block' width='1em' height='1em' xmlns='http://www.w3.org/2000/svg' %3E%3Cpath fill='none' stroke='currentColor' stroke-linecap='round' stroke-linejoin='round' d='M9.5 16.5v-4.978l-5-.022m14-9l-7 7m5 0l-5 .023V4.5m-2 7l-7 7'/%3E%3C/svg%3E");-webkit-mask:var(--un-icon) no-repeat;mask:var(--un-icon) no-repeat;-webkit-mask-size:100% 100%;mask-size:100% 100%;background-color:currentColor;color:inherit;display:inline-block;width:1em;height:1em;}
/* layer: basestyles */

          html,
          body,
          #app {
            height: 100%;
            margin: 0;
            padding: 0;
          }

          html.dark {
            background: #121212;
            color-scheme: dark;
          }

          :root {
            color-scheme: light dark;
            color: rgba(255, 255, 255, 0.87);
            background-color: #242424;
          }

          @media (prefers-color-scheme: light) {
            :root {
              color: #213547;
              background-color: #ffffff;
            }
            a:hover {
              color: #747bff;
            }
          }
        
/* layer: default */
.\\[--row-gap\\:3px\\]{--row-gap:3px;}
.bg-primary{--un-bg-opacity:1;background-color:rgb(255 87 51 / var(--un-bg-opacity));}
.text-sm{font-size:0.875rem;line-height:1.25rem;}
.text-size-1{font-size:0.25rem;}"
`;
